src/unistd.rs

```diff
@@ -2136,3 +2146,424 @@ mod setres {
+/// // group.
+/// let res = Group::query( GroupQuery::Name(find.clone()) ).unwrap().unwrap();
+/// assert!(res.name == find);
+/// ```
+#[derive(Debug)]
+pub enum GroupQuery {
+    /// Get a group by GID. 
+    ///
+    /// Internally, this function calls
+    /// [getgrgid_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Gid(Gid),
+    /// Get a group by name. 
+    ///
+    /// Internally, this function calls
+    /// [getgrnam_r(3)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/getpwuid_r.html)
+    Name(String),
+    GidWithBufsize(Gid, usize),
+    NameWithBufsize(String, usize)
+}
+
+pub trait Queryable<Q> where Self: ::std::marker::Sized {
+    fn query(q: Q) -> Option<Result<Self>>;
+}
+
+/// Default buffer size for system user and group querying functions
+const PWGRP_BUFSIZE: usize = 1024;
+
+impl Queryable<UserQuery> for User {
+    fn query(q: UserQuery) -> Option<Result<Self>> {
+        let mut cbuf = match q {
+            UserQuery::UidWithBufsize(_, size) | 
+            UserQuery::NameWithBufsize(_, size) => {
+                vec![0 as c_char; size]
+            },
+            _ => {
+                vec![0 as c_char; PWGRP_BUFSIZE]
+            }
+        };
+
+        let mut pwd: libc::passwd =  unsafe { mem::zeroed() };
+        let mut res = ptr::null_mut();
+
+        let i = { 
+            unsafe { Errno::clear(); }
+
+            match q {
+                UserQuery::UidWithBufsize(Uid(uid), _) |
+                UserQuery::Uid(Uid(uid)) => {
+                    unsafe {
+                        libc::getpwuid_r(uid, &mut pwd,
+                                         cbuf.as_mut_ptr(),
+                                         cbuf.len(), &mut res) 
+                    }
+                },
+                UserQuery::NameWithBufsize(name, _) |
+                UserQuery::Name(name) => {
+                    unsafe {
+                        libc::getpwnam_r(CString::new(name).unwrap().as_ptr(), 
+                                         &mut pwd, cbuf.as_mut_ptr(), cbuf.len(), &mut res) 
+                    }
+                },
+            }
+
+        };
+
+        match i {
+            0 => {
+                if !res.is_null() {
+                    Some(Ok(User::from(res))) 
+                } else {
+                    None
+                }
+            },
+            _ => { Some(Err(Error::Sys(Errno::last()))) },
+        }
+    }
+}
+
+impl Queryable<GroupQuery> for Group {
+    fn query(q: GroupQuery) -> Option<Result<Self>> {
+        let mut cbuf = match q {
+            GroupQuery::GidWithBufsize(_, size) | 
+            GroupQuery::NameWithBufsize(_, size) => {
+                vec![0 as c_char; size]
+            },
+            _ => {
+                vec![0 as c_char; PWGRP_BUFSIZE]
+            }
+        };
+
+        let mut grp: libc::group =  unsafe { mem::zeroed() };
+        let mut res = ptr::null_mut();
+
+        let i = {
```

Comment:
Same here, please don't use `i`, a more descriptive variable name is apprciated.